modules/rgbd/test/test_tsdf.cpp

```diff
@@ -0,0 +1,256 @@
+            Affine3f pose;
+            pose = pose.rotate(startPose.rotation());
+            pose = pose.rotate(Vec3f(0.f, -1.f, 0.f) * angle);
+            pose = pose.translate(Vec3f(startPose.translation()[0] * sin(angle),
+                startPose.translation()[1],
+                startPose.translation()[2] * cos(angle)));
+            poses.push_back(pose);
+        }
+
+        return poses;
+    }
+
+};
+
+Ptr<Scene> Scene::create(int nScene, Size sz, Matx33f _intr, float _depthFactor)
+{
+    return makePtr<SemisphereScene>(sz, _intr, _depthFactor);
+}
+
+struct Operator {
+    void operator ()(Vec4f& vector, const int* position) const
+    {
+        if ( !isnan(vector[0]) )
+        {
+            float length = vector[0] * vector[0] +
+                           vector[1] * vector[1] +
+                           vector[2] * vector[2];
+            //cout << length;
+            ASSERT_TRUE(0.9999f < length && length < 1.0001f);
+        }
+    }
+};
+
+static const bool display = false;
+
+void normal_test(bool hiDense, bool isHashTSDF)
+{
+    Ptr<kinfu::Params> _params;
+    if (isHashTSDF)
+        _params = kinfu::Params::hashTSDFParams(true);
+    else
+        _params = kinfu::Params::coarseParams();
+
+    Ptr<Scene> scene = Scene::create(hiDense, _params->frameSize, _params->intr, _params->depthFactor);
+    std::vector<Affine3f> poses = scene->getPoses();
+
+    Mat depth = scene->depth(poses[0]);
+    UMat _points, _normals;
+
+    Ptr<kinfu::Volume> volume = kinfu::makeVolume(_params->volumeType, _params->voxelSize, _params->volumePose, 
+                                _params->raycast_step_factor, _params->tsdf_trunc_dist, _params->tsdf_max_weight, 
+                                _params->truncateThreshold, _params->volumeDims);
+
+    volume->integrate(depth, _params->depthFactor, poses[0], _params->intr);
+    volume->raycast(poses[0], _params->intr, _params->frameSize, _points, _normals);
+    
+    //volume->fetchPointsNormals(_points, _normals);
+    //volume->fetchNormals(_points, _normals);
+ 
+    AccessFlag af = ACCESS_READ;
+    Mat normals = _normals.getMat(af);
+    normals.forEach<Vec4f>(Operator());
+
+    if (display)
+    {
+        imshow("depth", depth * (1.f / _params->depthFactor / 4.f));
+        Mat points = _points.getMat(af);
+        Mat image;
+        //renderPointsNormals(points, normals, image, _params->lightPose);
+        imshow("render", image);
+        waitKey(30000);
+    }
+
+    UMat _newPoints, _newNormals;
+    volume->raycast(poses[17], _params->intr, _params->frameSize, _newPoints, _newNormals);
+
+    normals = _newNormals.getMat(af);
+    normals.forEach<Vec4f>(Operator());
+
+    if (display)
+    {
+        imshow("depth", depth * (1.f / _params->depthFactor / 4.f));
+        Mat points = _newPoints.getMat(af);
+        Mat image;
+        //renderPointsNormals(points, normals, image, _params->lightPose);
+        imshow("render", image);
+        waitKey(30000);
+    }
```

Comment:
Here valid points should be counted and checked. `countNonZeros` can be used for that.

Reply:
done